### ISSUE TITLE
Discrepancy in Outputs When Converting to ONNX using torch.autocast with float16

### ISSUE THREAD

**Issue by AyoubMDL** [pytorch/pytorch#146312]
### 🐛 Describe the bug

I converted a PyTorch model to ONNX inside `torch.autocast(device_type="cpu", dtype=torch.float16)` context. When running inference on both the original and converted models, I observed that the difference between their outputs is on the order of **1e-2**.

I expected a smaller error, around 1e-5, similar to float32 precision, but I assume that in float16, the error might naturally be higher.

### Questions:
* Is this level of discrepancy expected for mixed-precision models eventhough it is just a converted model (no added ops) ?
* Are there any recommended ways to minimize the difference between PyTorch and ONNX Runtime outputs in FLOAT16 (I believe they will always be an error because of the optimization differences) ?

#### Additional Notes:
Upgrading to the latest PyTorch version (which has FP16 support on CPU) reduced the error but it remains in the 1e-2 range.
Would appreciate any insights on this. Thanks!

### Versions

PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.31.3
Libc version: glibc-2.31

Python version: 3.11.11 (main, Dec  4 2024, 08:55:08) [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-130-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080 Ti
Nvidia driver version: 550.107.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Little Endian
Address sizes:                        39 bits physical, 48 bits virtual
CPU(s):                               16
On-line CPU(s) list:                  0-15
Thread(s) per core:                   2
Core(s) per socket:                   8
Socket(s):                            1
NUMA node(s):                         1
Vendor ID:                            GenuineIntel
CPU family:                           6
Model:                                165
Model name:                           Intel(R) Core(TM) i7-10700 CPU @ 2.90GHz
Stepping:                             5
CPU MHz:                              4187.537
CPU max MHz:                          4800,0000
CPU min MHz:                          800,0000
BogoMIPS:                             5799.77
Virtualization:                       VT-x
L1d cache:                            256 KiB
L1i cache:                            256 KiB
L2 cache:                             2 MiB
L3 cache:                             16 MiB
NUMA node0 CPU(s):                    0-15
Vulnerability Gather data sampling:   Mitigation; Microcode
Vulnerability Itlb multihit:          KVM: Mitigation: VMX disabled
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Mitigation; Microcode
Vulnerability Tsx async abort:        Not affected
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp pku ospke md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] onnx==1.16.1
[pip3] onnxconverter-common==1.16.0
[pip3] onnxmltools==1.13.0
[pip3] onnxruntime==1.19.0
[pip3] onnxruntime_extensions==0.13.0
[pip3] onnxscript==0.1.0.dev20250114
[pip3] pytorch-lightning==1.9.5
[pip3] torch==2.6.0
[pip3] torchao==0.8.0
[pip3] torchaudio==2.5.1
[pip3] torchmetrics==1.6.1
[pip3] torchtune==0.5.0
[pip3] torchvision==0.20.1
[pip3] triton==3.2.0
[conda] Could not collect

**Comment by justinchuby** [comment]
How did you export the model? Can you share a snippet? The discrepancies may vary depending on the operators you have in your model

**Comment by justinchuby** [comment]
If you are running onnx runtime, it does not have float16 kernels for CPU. Float16 is supported by adding casts to float32 on cpu. So that may contribute to some errors. You need to compare the gpu results (or other supported execution providers) for float16 models.

**Comment by AyoubMDL** [comment]
Hi @justinchuby , thanks for the response. Here's a snippet of the onnx conversion
```python
def convert_to_onnx(torch_model, save_path, external_data=True):
    torch_model = torch_model.to("cpu")
    dummy_inputs = torch.randint(128, size=(1,), dtype=torch.int32).to("cpu")

    # Run forward to build some layers
    with torch.no_grad():
        torch_model(dummy_inputs)

    # Export to onnx
    with tempfile.TemporaryDirectory() as tmp_dir:
        tmp_path = os.path.join(tmp_dir, "tmp.onnx")
        with torch.no_grad(), torch.autocast(device_type="cpu", dtype=torch.float16):
            onnx_program = torch.onnx.export(torch_model,
                                             dummy_inputs,
                                             tmp_path,
                                             input_names=["input"],
                                             opset_version=15,
                                             external_data=external_data,
                                             dynamo=True)

            onnx_program.optimize()
            onnx_program.save(tmp_path, external_data=external_data)

        # ONNXRuntime optimization
        _optimize_onnx_model(tmp_path, save_path)
```
My goal is to obtain an ONNX model that closely matches the PyTorch model, which should be feasible since it's a conversion process.

1. I use the `torch.autocast` context to allow ONNX export to trace type casts, helping me identify which operations are performed in float16.
2. Additionally, I noticed that the ONNX model stores weights in float16 but casts them back to float32. I’m wondering whether the PyTorch model follows the same pattern—casting weights to float16 before converting them back to float32 for computations.

Below is a comparison of discrepancies between PyTorch and ONNXRuntime on both CPU and GPU. The GPU error is slightly higher, likely due to custom nodes running on the CPU:
|   | Max Absolute Difference (Pytorch VS ONNXRuntime) | Mean Absolute Difference (Pytorch VS ONNXRuntime) |
|---|-----------------------------------------------|-----------------------------------------------|
| CPU | 0.0195312 | 0.00193405 |
| GPU | 0.0224609 | 0.0024147 |

In FLOAT32 conversion, I have difference on the order of 1e-5 which is acceptable.
In your opinion, are these differences acceptable for FLOAT16/Mixed precision ? or is there something unusual here?
Tell me if you need additional information.

**Comment by justinchuby** [comment]
This error is typically considered reasonable, but that is dependent on your use case. How big is your model? You can set report=True in the export call and attach the markdown report for us to take a look.

**Comment by AyoubMDL** [comment]
The model has approximately 1.7B parameters (it's a mamba-like model). For instance, when I evaluate the MMLU, I observe a difference of 0.02. While this difference isn't significant, I believe that when exporting the model, I shouldn't expect to see this kind of discrepancy (I don't have such issue when I work with the float32 model).
Here is the markdown report: 

[onnx_export_2025-02-04_10-55-03-647304_success.md](https://github.com/user-attachments/files/18655547/onnx_export_2025-02-04_10-55-03-647304_success.md)

**Comment by justinchuby** [comment]
I think the discrepancy is within reason. Since ONNX and PyTorch does not have the same set of operators or the same implementation, a PyTorch operator may be converted and represented using a composition of ONNX operators, which could introduce precision loss. The difference in runtime implementation can introduce numerical differences too (which can mean increased or reduced precision)

**Comment by AyoubMDL** [comment]
Got it, I have two follow-up questions:

1. In the ONNX model, almost all operations are performed in float32. According to [PyTorch’s AMP documentation](https://pytorch.org/docs/stable/amp.html#cpu-ops-that-can-autocast-to-bfloat16), operations like linear and matmul can be autocast to float16. Is it because casting them to float16 would reduce accuracy, so autocast keeps them in float32?
2. The ONNX weights are stored in float16 but are cast to float32 during inference. Is this also the case in PyTorch, or does PyTorch directly use float32 weights without any casting?

**Comment by justinchuby** [comment]
1. We use torch.export decomposition to handle autocast. So ONNX exporter itself does not make any decisions about autocast
2. The conversion logic translates the graph as is. I assume that’s what PyTorch eager does too. But I don’t know for sure.

**Comment by AyoubMDL** [comment]
Thank you for the responses ! Closing this issue.
